### PR TITLE
python3Packages.treeo: 0.4.0 -> 0.0.9

### LINKED ISSUE #159455: `python3Packages.treex` build failure on x86_64-linux as of `70aa8d9e`
## Issue description
Build of `python3Packages.treex` failed on x86_64-linux as of 70aa8d9ec6e80e9d6e5967c552ce665859863303.

[build log](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1832991923)

Maintainer cc: @ndl
Other cc: n/a

### Steps to reproduce
1. Checkout nixpkgs at commit 70aa8d9ec6e80e9d6e5967c552ce665859863303
2. Run `nix-build -A python3Packages.treex`

## Technical details
```
- system: `"x86_64-linux"`
 - host os: `Linux 5.11.0-1028-azure, Ubuntu, 20.04.3 LTS (Focal Fossa)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.0`
 - nixpkgs: `/nix/store/f78jqkss0z39h2602xrrigzy6g1skpv6-source`
```

## Misc.
This issue was automatically generated by [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep).
- [CI workflow](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1832991923) that created this issue.
- Internal tag: 8597461721032410299

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
fixes https://github.com/NixOS/nixpkgs/issues/159455

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 159933` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python39Packages.elegy</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python310Packages.treeo</li>
    <li>python39Packages.treeo</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

File: pkgs/development/python-modules/treeo/default.nix
Comment (samuela) [inline pkgs/development/python-modules/treeo/default.nix:29]: Thank you! This should not have been in `propagatedBuildInputs` to begin with per https://nixos.wiki/wiki/JAX#How_do_I_package_JAX_libraries.3F.

File: pkgs/development/python-modules/treeo/default.nix
Comment (samuela) [inline pkgs/development/python-modules/treeo/default.nix:12]: From a semver perspective this is a downgrade. Why should we downgrade instead of upgrading?

**Review by samuela — CHANGES_REQUESTED** [review]
just confused re the version downgrade. o/w LGTM. if the 0.0.9 version is actually newer than 0.4.0, I think it my actually be preferable to just move to an `unstable-yyyy-mm-dd` version.

**Comment by samuela** [comment]
elegy build failure is unrelated and was already broken

**Comment by dotlambda** [comment]
> just confused re the version downgrade. o/w LGTM. if the 0.0.9 version is actually newer than 0.4.0, I think it my actually be preferable to just move to an `unstable-yyyy-mm-dd` version.

The previous version was accidentally called 0.4.0 instead of 0.0.4. See https://github.com/cgarciae/treeo/releases and https://pypi.org/project/treeo/#history.
So 0.0.9 is the most recent version.
I don't see why we should deviate from upstream versioning.

**Comment by samuela** [comment]
Maybe I'm missing something but I don't see any indication that 0.4.0 was an accident in the release notes. Is there a reference for this?

**Comment by dotlambda** [comment]
> Maybe I'm missing something but I don't see any indication that 0.4.0 was an accident in the release notes. Is there a reference for this?

Compare the version history on PyPI with the one on GitHub.

**Comment by samuela** [comment]
I'm a little confused so I'll defer to maintainer @ndl on this one.

File: pkgs/development/python-modules/treeo/default.nix
Comment (ndl) [reply pkgs/development/python-modules/treeo/default.nix:29]: Wait, isn't that the opposite of what we should be doing here? IIUC it was correct before the change (= `jaxlib` used only in `checkInputs` and not in `propagatedBuildInputs`).

@dotlambda please keep it the way it was before (= _not_ propagating `jaxlib` and explicitly passing it in `checkInputs`) as otherwise it forces the specific version of `jaxlib` onto the user, see the link above for details.

File: pkgs/development/python-modules/treex/default.nix
Comment (ndl) [inline pkgs/development/python-modules/treex/default.nix:42]: Same as above - please keep it the way it is.

File: pkgs/development/python-modules/treeo/default.nix
Comment (ndl) [reply pkgs/development/python-modules/treeo/default.nix:12]: I see the point @dotlambda makes, it looks like upstream versioning failure indeed.

**Review by ndl — CHANGES_REQUESTED** [review]
Using `0.0.9` instead of `0.4.0` looks the right step, thanks for catching the issue!

The only issue is with propagating `jaxlib`, please revert that part and I'll approve the rest.

File: pkgs/development/python-modules/treeo/default.nix
Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix:29]: oops! I was way too tired when I wrote that! yes jaxlib should absolutely _not_ be in `propagatedBuildInputs`. Thanks for catching @ndl!

Comment (SuperSandro2000) [reply pkgs/development/python-modules/treeo/default.nix:29]: Normally we use upstreams requirements as a good recommendations but many python packages use the wrong requires or force linters we don't want. If that package is required for this package to work correct then it can be propagated even if upstream has done something different.

Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix:29]: We have an established policy of not propagating `jaxlib` in the JAX ecosystem. Sometimes that means deviating from upstreams dependency lists, and creating issues upstream as well.

**Comment by smancill** [comment]
Done by db60f460b67463cdc12c3ad2982a49a80269717e

**Pull request closed without merging**
